{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "reverse-cooperative",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import sqlite3\n",
    "from collections import Counter\n",
    "\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "unknown-representation",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "# remove test.db and ignore the error if it does not exist\n",
    "rm test.db | exit 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "pending-original",
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_ROOT = \"../exascale_data\"\n",
    "con = sqlite3.connect(\"test.db\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "numeric-spoke",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['index',\n",
       " 'node_id',\n",
       " 'node_type',\n",
       " 'gene_model_type',\n",
       " 'gene_symbol',\n",
       " 'go_description',\n",
       " 'go_terms',\n",
       " 'mapman_bin',\n",
       " 'mapman_description',\n",
       " 'mapman_name',\n",
       " 'pheno_aragwas_id',\n",
       " 'pheno_description',\n",
       " 'pheno_pto_description',\n",
       " 'pheno_pto_name',\n",
       " 'pheno_reference',\n",
       " 'tair_computational_description',\n",
       " 'tair_curator_summary',\n",
       " 'tair_short_description',\n",
       " 'transcript']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clusters_table = \"\"\"CREATE TABLE IF NOT EXISTS clusters (\n",
    "    \"index\" INTEGER,\n",
    "    \"cluster_id\" TEXT,\n",
    "    \"cluster_prefix\" TEXT,\n",
    "    \"node_ids\" TEXT\n",
    ")\"\"\"\n",
    "\n",
    "edges_table = \"\"\"CREATE TABLE IF NOT EXISTS edges (\n",
    "    \"index\" INTEGER,\n",
    "    \"node1\" TEXT,\n",
    "    \"node2\" TEXT,\n",
    "    \"score\" REAL,\n",
    "    \"edge_type\" TEXT,\n",
    "    \"directed\" INTEGER\n",
    ")\"\"\"\n",
    "\n",
    "nodes_raw_table = \"\"\"CREATE TABLE IF NOT EXISTS nodes_raw (\n",
    "    \"index\" INTEGER,\n",
    "    \"node_id\" TEXT,\n",
    "    \"node_type\" TEXT,\n",
    "    \"gene_model_type\" TEXT,\n",
    "    \"gene_symbol\" TEXT,\n",
    "    \"go_description\" TEXT,\n",
    "    \"go_terms\" TEXT,\n",
    "    \"mapman_bin\" TEXT,\n",
    "    \"mapman_description\" TEXT,\n",
    "    \"mapman_name\" TEXT,\n",
    "    \"pheno_aragwas_id\" TEXT,\n",
    "    \"pheno_description\" TEXT,\n",
    "    \"pheno_pto_description\" TEXT,\n",
    "    \"pheno_pto_name\" TEXT,\n",
    "    \"pheno_reference\" TEXT,\n",
    "    \"tair_computational_description\" TEXT,\n",
    "    \"tair_curator_summary\" TEXT,\n",
    "    \"tair_short_description\" TEXT,\n",
    "    \"transcript\" TEXT\n",
    ")\"\"\"\n",
    "\n",
    "nodes_table = \"\"\"CREATE TABLE IF NOT EXISTS nodes (\n",
    "    \"mindex\" INTEGER,\n",
    "    \"node_id\" TEXT,\n",
    "    \"node_type\" TEXT,\n",
    "    \"gene_model_type\" TEXT,\n",
    "    \"gene_symbol\" TEXT,\n",
    "    \"pheno_aragwas_id\" TEXT,\n",
    "    \"pheno_description\" TEXT,\n",
    "    \"pheno_pto_description\" TEXT,\n",
    "    \"pheno_pto_name\" TEXT,\n",
    "    \"pheno_reference\" TEXT,\n",
    "    \"transcripts\" TEXT,\n",
    "    \"go_terms\" TEXT,\n",
    "    \"go_infos\" TEXT,\n",
    "    \"mapman_infos\" TEXT,\n",
    "    \"tair_infos\" TEXT\n",
    ")\"\"\"\n",
    "\n",
    "cur_clusters = con.execute(clusters_table)\n",
    "cur_edges = con.execute(edges_table)\n",
    "cur_nodes_raw = con.execute(nodes_raw_table)\n",
    "cur_nodes = con.execute(nodes_table)\n",
    "\n",
    "assert cur_clusters.fetchall() == []\n",
    "assert cur_edges.fetchall() == []\n",
    "assert cur_nodes_raw.fetchall() == []\n",
    "assert cur_nodes.fetchall() == []\n",
    "\n",
    "columns_nodes_raw = list(pd.read_sql(\"\"\"SELECT * FROM nodes_raw\"\"\", con).columns)\n",
    "columns_nodes_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "separated-flood",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>score</th>\n",
       "      <th>edge_type</th>\n",
       "      <th>directed</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AT2G18790</td>\n",
       "      <td>AT1G09530</td>\n",
       "      <td>1.000</td>\n",
       "      <td>protein-protein-interaction_Mentha_A_thaliana_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AT4G33430</td>\n",
       "      <td>AT4G39400</td>\n",
       "      <td>1.000</td>\n",
       "      <td>protein-protein-interaction_Mentha_A_thaliana_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AT3G20740</td>\n",
       "      <td>AT1G02580</td>\n",
       "      <td>0.999</td>\n",
       "      <td>protein-protein-interaction_Mentha_A_thaliana_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AT5G35410</td>\n",
       "      <td>AT4G33000</td>\n",
       "      <td>0.999</td>\n",
       "      <td>protein-protein-interaction_Mentha_A_thaliana_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AT3G62980</td>\n",
       "      <td>AT3G23050</td>\n",
       "      <td>0.999</td>\n",
       "      <td>protein-protein-interaction_Mentha_A_thaliana_...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node1      node2  score  \\\n",
       "0  AT2G18790  AT1G09530  1.000   \n",
       "1  AT4G33430  AT4G39400  1.000   \n",
       "2  AT3G20740  AT1G02580  0.999   \n",
       "3  AT5G35410  AT4G33000  0.999   \n",
       "4  AT3G62980  AT3G23050  0.999   \n",
       "\n",
       "                                           edge_type  directed  \n",
       "0  protein-protein-interaction_Mentha_A_thaliana_...         1  \n",
       "1  protein-protein-interaction_Mentha_A_thaliana_...         1  \n",
       "2  protein-protein-interaction_Mentha_A_thaliana_...         1  \n",
       "3  protein-protein-interaction_Mentha_A_thaliana_...         1  \n",
       "4  protein-protein-interaction_Mentha_A_thaliana_...         1  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1 = pd.read_csv(f\"{DATA_ROOT}/prerelease/edge_data/mentha_AT-PPI-3702_040319.to-kbase-spec.tsv\", sep=\"\\t\")\n",
    "df1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "consolidated-valentine",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node1</th>\n",
       "      <th>node2</th>\n",
       "      <th>score</th>\n",
       "      <th>edge_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AT5G67420</td>\n",
       "      <td>AT4G09820</td>\n",
       "      <td>1</td>\n",
       "      <td>transcription-factor-regulatory-interaction_li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AT2G20180</td>\n",
       "      <td>AT4G23750</td>\n",
       "      <td>1</td>\n",
       "      <td>transcription-factor-regulatory-interaction_li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AT2G20180</td>\n",
       "      <td>AT1G80340</td>\n",
       "      <td>1</td>\n",
       "      <td>transcription-factor-regulatory-interaction_li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AT2G20180</td>\n",
       "      <td>AT2G29090</td>\n",
       "      <td>1</td>\n",
       "      <td>transcription-factor-regulatory-interaction_li...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AT2G20180</td>\n",
       "      <td>AT2G40220</td>\n",
       "      <td>1</td>\n",
       "      <td>transcription-factor-regulatory-interaction_li...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       node1      node2  score  \\\n",
       "0  AT5G67420  AT4G09820      1   \n",
       "1  AT2G20180  AT4G23750      1   \n",
       "2  AT2G20180  AT1G80340      1   \n",
       "3  AT2G20180  AT2G29090      1   \n",
       "4  AT2G20180  AT2G40220      1   \n",
       "\n",
       "                                           edge_type  \n",
       "0  transcription-factor-regulatory-interaction_li...  \n",
       "1  transcription-factor-regulatory-interaction_li...  \n",
       "2  transcription-factor-regulatory-interaction_li...  \n",
       "3  transcription-factor-regulatory-interaction_li...  \n",
       "4  transcription-factor-regulatory-interaction_li...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df0 = pd.read_csv(f\"{DATA_ROOT}/prerelease/edge_data/ATRM_TF_to_Target_LitCurated_01082020.to-kbase-spec.tsv\", sep=\"\\t\")\n",
    "df0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "concerned-package",
   "metadata": {},
   "outputs": [],
   "source": [
    "df0.to_sql(\"edges\", con, if_exists=\"append\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "tropical-tenant",
   "metadata": {},
   "outputs": [],
   "source": [
    "df1.to_sql(\"edges\", con, if_exists=\"append\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "macro-portugal",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>letter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>User 4</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>User 5</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>User 6</td>\n",
       "      <td>a</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name letter\n",
       "0  User 4      a\n",
       "1  User 5      a\n",
       "2  User 6      a"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2 = pd.DataFrame({'name' : ['User 4', 'User 5', 'User 6']})\n",
    "df3 = pd.DataFrame(dict(letter=[\"a\"]*len(df2)))\n",
    "pd.concat([df2, df3], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dried-israel",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>node_ids</th>\n",
       "      <th>cluster_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3097</td>\n",
       "      <td>Cluster3098</td>\n",
       "      <td>AT2G20080,AT2G20660</td>\n",
       "      <td>markov_i2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3098</td>\n",
       "      <td>Cluster3099</td>\n",
       "      <td>AT3G62280,AT5G23210</td>\n",
       "      <td>markov_i2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3099</td>\n",
       "      <td>Cluster3100</td>\n",
       "      <td>AT1G75890,AT3G15630</td>\n",
       "      <td>markov_i2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3100</td>\n",
       "      <td>Cluster3101</td>\n",
       "      <td>AT1G67670,AT1G78710</td>\n",
       "      <td>markov_i2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3101</td>\n",
       "      <td>Cluster3102</td>\n",
       "      <td>AT1G54870,AT5G54740</td>\n",
       "      <td>markov_i2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id             node_ids cluster_prefix\n",
       "3097  Cluster3098  AT2G20080,AT2G20660      markov_i2\n",
       "3098  Cluster3099  AT3G62280,AT5G23210      markov_i2\n",
       "3099  Cluster3100  AT1G75890,AT3G15630      markov_i2\n",
       "3100  Cluster3101  AT1G67670,AT1G78710      markov_i2\n",
       "3101  Cluster3102  AT1G54870,AT5G54740      markov_i2"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df5 = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/cluster_data/out.aranetv2_subnet_AT-CX_top10percent_anno_AF_082919.abc.I2_named.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "df6 = pd.DataFrame(dict(cluster_prefix=[\"markov_i2\"]*len(df5))) # i2 - inflation of 2\n",
    "df7 = pd.concat([df5, df6], axis=1)\n",
    "df7.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "indian-paintball",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>cluster_prefix</th>\n",
       "      <th>node_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [index, cluster_id, cluster_prefix, node_ids]\n",
       "Index: []"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"select * from clusters order by \"index\" desc limit 5\"\"\", con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "peaceful-chest",
   "metadata": {},
   "outputs": [],
   "source": [
    "df7.to_sql(\"clusters\", con, if_exists=\"append\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "limited-actress",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>node_ids</th>\n",
       "      <th>cluster_prefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>3688</td>\n",
       "      <td>Cluster3689</td>\n",
       "      <td>AT2G20080,AT2G20660</td>\n",
       "      <td>markov_i4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3689</td>\n",
       "      <td>Cluster3690</td>\n",
       "      <td>AT3G62280,AT5G23210</td>\n",
       "      <td>markov_i4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3690</td>\n",
       "      <td>Cluster3691</td>\n",
       "      <td>AT1G75890,AT3G15630</td>\n",
       "      <td>markov_i4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3691</td>\n",
       "      <td>Cluster3692</td>\n",
       "      <td>AT1G67670,AT1G78710</td>\n",
       "      <td>markov_i4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3692</td>\n",
       "      <td>Cluster3693</td>\n",
       "      <td>AT1G54870,AT5G54740</td>\n",
       "      <td>markov_i4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       cluster_id             node_ids cluster_prefix\n",
       "3688  Cluster3689  AT2G20080,AT2G20660      markov_i4\n",
       "3689  Cluster3690  AT3G62280,AT5G23210      markov_i4\n",
       "3690  Cluster3691  AT1G75890,AT3G15630      markov_i4\n",
       "3691  Cluster3692  AT1G67670,AT1G78710      markov_i4\n",
       "3692  Cluster3693  AT1G54870,AT5G54740      markov_i4"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df8 = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/cluster_data/out.aranetv2_subnet_AT-CX_top10percent_anno_AF_082919.abc.I4_named.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "df9 = pd.DataFrame(dict(cluster_prefix=[\"markov_i4\"]*len(df8))) # i4 - inflation of 4\n",
    "dfx = pd.concat([df8, df9], axis=1)\n",
    "dfx.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "chemical-korea",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfx.to_sql(\"clusters\", con, if_exists=\"append\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "provincial-bermuda",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>cluster_id</th>\n",
       "      <th>cluster_prefix</th>\n",
       "      <th>node_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>3692</td>\n",
       "      <td>Cluster3693</td>\n",
       "      <td>markov_i4</td>\n",
       "      <td>AT1G54870,AT5G54740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>3691</td>\n",
       "      <td>Cluster3692</td>\n",
       "      <td>markov_i4</td>\n",
       "      <td>AT1G67670,AT1G78710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3690</td>\n",
       "      <td>Cluster3691</td>\n",
       "      <td>markov_i4</td>\n",
       "      <td>AT1G75890,AT3G15630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3689</td>\n",
       "      <td>Cluster3690</td>\n",
       "      <td>markov_i4</td>\n",
       "      <td>AT3G62280,AT5G23210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>3688</td>\n",
       "      <td>Cluster3689</td>\n",
       "      <td>markov_i4</td>\n",
       "      <td>AT2G20080,AT2G20660</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index   cluster_id cluster_prefix             node_ids\n",
       "0   3692  Cluster3693      markov_i4  AT1G54870,AT5G54740\n",
       "1   3691  Cluster3692      markov_i4  AT1G67670,AT1G78710\n",
       "2   3690  Cluster3691      markov_i4  AT1G75890,AT3G15630\n",
       "3   3689  Cluster3690      markov_i4  AT3G62280,AT5G23210\n",
       "4   3688  Cluster3689      markov_i4  AT2G20080,AT2G20660"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(\"\"\"select * from clusters order by \"index\" desc limit 5\"\"\", con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "04019db3-af0c-428b-829d-bcc9bcdcb9b1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "==> ../exascale_data/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_gene-GOslimTAIR-v010121.tsv <==\n",
      "node_id\tgo_description\tgo_terms\ttranscript\tnode_type\n",
      "\n",
      "==> ../exascale_data/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_pheno-AraPheno-v062716.tsv <==\n",
      "node_id\tnode_type\tpheno_aragwas_id\tpheno_description\tpheno_pto_name\tpheno_pto_description\tpheno_reference\n",
      "\n",
      "==> ../exascale_data/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_gene-MapMan-vX4.2.tsv <==\n",
      "mapman_bin\tmapman_name\ttranscript\tmapman_description\tnode_id\tnode_type\n",
      "\n",
      "==> ../exascale_data/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_gene-TAIR-v2018.tsv <==\n",
      "node_id\ttranscript\tgene_model_type\ttair_short_description\ttair_curator_summary\ttair_computational_description\tnode_type\n"
     ]
    }
   ],
   "source": [
    "# columns for each node source file\n",
    "!find {DATA_ROOT}/prerelease/node_tables -maxdepth 1 -mindepth 1 | xargs head -n 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "greater-function",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfn0 = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_gene-MapMan-vX4.2.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "dfn1 = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_gene-TAIR-v2018.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "dfn2 = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_pheno-AraPheno-v062716.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "dfn3 = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/node_tables/20210107_ExaKBase_AT_nodetab_gene-GOslimTAIR-v010121.tsv\",\n",
    "    sep=\"\\t\"\n",
    ")\n",
    "merge_df = pd.read_csv(\n",
    "    f\"{DATA_ROOT}/prerelease/aranet2-aragwas-MERGED-AMW-v2_091319_nodeTable.csv\",\n",
    "    usecols=lambda col: col in columns_nodes_raw\n",
    ")\n",
    "dfn0.to_sql(\"nodes_raw\", con, if_exists=\"append\")\n",
    "dfn1.to_sql(\"nodes_raw\", con, if_exists=\"append\")\n",
    "dfn2.to_sql(\"nodes_raw\", con, if_exists=\"append\")\n",
    "dfn3.to_sql(\"nodes_raw\", con, if_exists=\"append\")\n",
    "merge_df.to_sql(\"nodes_raw\", con, if_exists=\"append\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "advised-expense",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AT1G07890</td>\n",
       "      <td>gene</td>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AT4G30820</td>\n",
       "      <td>gene</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AT1G20620</td>\n",
       "      <td>gene</td>\n",
       "      <td>74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AT4G20260</td>\n",
       "      <td>gene</td>\n",
       "      <td>66</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AT1G51270</td>\n",
       "      <td>gene</td>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     node_id node_type  counts\n",
       "0  AT1G07890      gene      92\n",
       "1  AT4G30820      gene      90\n",
       "2  AT1G20620      gene      74\n",
       "3  AT4G20260      gene      66\n",
       "4  AT1G51270      gene      61"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# How many rows per node are there? If counts > 1 then some sort of merging is required.\n",
    "pd.read_sql(\"\"\"\n",
    "SELECT node_id, node_type, COUNT(*) as counts\n",
    "    FROM nodes_raw\n",
    "    GROUP BY node_id, node_type\n",
    "    ORDER BY count(*) DESC\n",
    "    LIMIT 5\n",
    "\"\"\", con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "impossible-carrier",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>counts</th>\n",
       "      <th>node_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>node_id</td>\n",
       "      <td>1</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>node_type</td>\n",
       "      <td>1</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>gene_model_type</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>gene_symbol</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>go_description</td>\n",
       "      <td>53</td>\n",
       "      <td>AT4G08920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>go_terms</td>\n",
       "      <td>52</td>\n",
       "      <td>AT4G08920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_bin</td>\n",
       "      <td>5</td>\n",
       "      <td>AT3G17240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_description</td>\n",
       "      <td>54</td>\n",
       "      <td>AT4G30820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_name</td>\n",
       "      <td>5</td>\n",
       "      <td>AT3G17240</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_aragwas_id</td>\n",
       "      <td>1</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_description</td>\n",
       "      <td>1</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_pto_description</td>\n",
       "      <td>1</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_pto_name</td>\n",
       "      <td>1</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_reference</td>\n",
       "      <td>2</td>\n",
       "      <td>SDV</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tair_computational_description</td>\n",
       "      <td>2</td>\n",
       "      <td>AT5G09710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tair_curator_summary</td>\n",
       "      <td>3</td>\n",
       "      <td>AT5G13220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tair_short_description</td>\n",
       "      <td>2</td>\n",
       "      <td>AT5G10140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>27</td>\n",
       "      <td>AT4G30820</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              key  counts    node_id\n",
       "0                         node_id       1        SDV\n",
       "0                       node_type       1        SDV\n",
       "0                 gene_model_type       1  ATMG01400\n",
       "0                     gene_symbol       1  ATMG01400\n",
       "0                  go_description      53  AT4G08920\n",
       "0                        go_terms      52  AT4G08920\n",
       "0                      mapman_bin       5  AT3G17240\n",
       "0              mapman_description      54  AT4G30820\n",
       "0                     mapman_name       5  AT3G17240\n",
       "0                pheno_aragwas_id       1        SDV\n",
       "0               pheno_description       1        SDV\n",
       "0           pheno_pto_description       1        SDV\n",
       "0                  pheno_pto_name       1        SDV\n",
       "0                 pheno_reference       2        SDV\n",
       "0  tair_computational_description       2  AT5G09710\n",
       "0            tair_curator_summary       3  AT5G13220\n",
       "0          tair_short_description       2  AT5G10140\n",
       "0                      transcript      27  AT4G30820"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# if counts <= 1 for each node_id then any represenative will do\n",
    "pd.concat([pd.read_sql(f\"\"\"\n",
    "SELECT '{key}' as key,\n",
    "        COUNT(DISTINCT({key})) as counts,\n",
    "        node_id\n",
    "    FROM nodes_raw\n",
    "    GROUP BY node_id\n",
    "    ORDER BY COUNT(DISTINCT({key})) DESC\n",
    "\"\"\", con).head(1) for key in columns_nodes_raw[1:]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "538ed25a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>counts</th>\n",
       "      <th>transcript</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>node_id</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>node_type</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>gene_model_type</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>gene_symbol</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>go_description</td>\n",
       "      <td>15</td>\n",
       "      <td>AT1G35720.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>go_terms</td>\n",
       "      <td>14</td>\n",
       "      <td>AT1G35720.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_bin</td>\n",
       "      <td>4</td>\n",
       "      <td>AT3G19710.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_description</td>\n",
       "      <td>4</td>\n",
       "      <td>AT3G17240.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_name</td>\n",
       "      <td>4</td>\n",
       "      <td>AT3G19710.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_aragwas_id</td>\n",
       "      <td>0</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_description</td>\n",
       "      <td>0</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_pto_description</td>\n",
       "      <td>0</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_pto_name</td>\n",
       "      <td>0</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>pheno_reference</td>\n",
       "      <td>0</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tair_computational_description</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tair_curator_summary</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01360.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>tair_short_description</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01360.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              key  counts   transcript\n",
       "0                         node_id       1  ATMG01400.1\n",
       "0                       node_type       1  ATMG01400.1\n",
       "0                 gene_model_type       1  ATMG01400.1\n",
       "0                     gene_symbol       1  ATMG01400.1\n",
       "0                  go_description      15  AT1G35720.1\n",
       "0                        go_terms      14  AT1G35720.1\n",
       "0                      mapman_bin       4  AT3G19710.1\n",
       "0              mapman_description       4  AT3G17240.3\n",
       "0                     mapman_name       4  AT3G19710.1\n",
       "0                pheno_aragwas_id       0  ATMG01400.1\n",
       "0               pheno_description       0  ATMG01400.1\n",
       "0           pheno_pto_description       0  ATMG01400.1\n",
       "0                  pheno_pto_name       0  ATMG01400.1\n",
       "0                 pheno_reference       0  ATMG01400.1\n",
       "0  tair_computational_description       1  ATMG01400.1\n",
       "0            tair_curator_summary       1  ATMG01360.1\n",
       "0          tair_short_description       1  ATMG01360.1\n",
       "0                      transcript       1  ATMG01400.1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([pd.read_sql(f\"\"\"\n",
    "SELECT '{key}' as key,\n",
    "        COUNT(DISTINCT({key})) as counts,\n",
    "        transcript\n",
    "    FROM nodes_raw\n",
    "    WHERE transcript IS NOT NULL\n",
    "    GROUP BY transcript\n",
    "    ORDER BY COUNT(DISTINCT({key})) DESC\n",
    "\"\"\", con).head(1) for key in columns_nodes_raw[1:]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "62e32a9e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>counts</th>\n",
       "      <th>mapman_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_bin</td>\n",
       "      <td>1</td>\n",
       "      <td>9.5.2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_description</td>\n",
       "      <td>9813</td>\n",
       "      <td>35.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>mapman_name</td>\n",
       "      <td>1</td>\n",
       "      <td>9.5.2.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  key  counts mapman_bin\n",
       "0          mapman_bin       1    9.5.2.5\n",
       "0  mapman_description    9813       35.1\n",
       "0         mapman_name       1    9.5.2.5"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.concat([pd.read_sql(f\"\"\"\n",
    "SELECT '{key}' as key,\n",
    "        COUNT(DISTINCT({key})) as counts,\n",
    "        mapman_bin\n",
    "    FROM nodes_raw\n",
    "    WHERE mapman_bin IS NOT NULL\n",
    "    GROUP BY mapman_bin\n",
    "    ORDER BY COUNT(DISTINCT({key})) DESC\n",
    "\"\"\", con).head(1) for key in (\"mapman_bin\", \"mapman_description\", \"mapman_name\")])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "af9505ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapman_bin</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>35.1</td>\n",
       "      <td>9813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>35.2</td>\n",
       "      <td>9261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>15.5.32</td>\n",
       "      <td>363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>50.1.13</td>\n",
       "      <td>354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>50.2.7</td>\n",
       "      <td>293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>95</td>\n",
       "      <td>15.5.3.5</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>96</td>\n",
       "      <td>15.5.1.2</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97</td>\n",
       "      <td>13.7.2.1</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98</td>\n",
       "      <td>50.3.6</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>99</td>\n",
       "      <td>18.4.7.4</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   mapman_bin  counts\n",
       "0        35.1    9813\n",
       "1        35.2    9261\n",
       "2     15.5.32     363\n",
       "3     50.1.13     354\n",
       "4      50.2.7     293\n",
       "..        ...     ...\n",
       "95   15.5.3.5      35\n",
       "96   15.5.1.2      35\n",
       "97   13.7.2.1      35\n",
       "98     50.3.6      34\n",
       "99   18.4.7.4      33\n",
       "\n",
       "[100 rows x 2 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_sql(f\"\"\"\n",
    "SELECT mapman_bin, COUNT(DISTINCT(mapman_description)) as counts\n",
    "    FROM nodes_raw\n",
    "    WHERE True\n",
    "        AND mapman_bin IS NOT NULL\n",
    "    GROUP BY mapman_bin\n",
    "    ORDER BY COUNT(DISTINCT(mapman_description)) DESC\n",
    "    LIMIT 100\n",
    "\"\"\", con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "19e01855",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(53, ('GO:0010343', 'GO:0043153', 'GO:2000652', 'GO:0009646', 'GO:0060918'))"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# go_terms has a many to one relationship with node_id\n",
    "go_terms_raw = tuple(frozenset(pd.read_sql(f\"\"\"\n",
    "SELECT node_id, go_terms\n",
    "    FROM nodes_raw\n",
    "    WHERE node_id = 'AT4G08920'\n",
    "\"\"\", con)[\"go_terms\"]))\n",
    "len(go_terms_raw), go_terms_raw[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e6b245ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(54,\n",
       " ('protein binding',\n",
       "  'flavin adenine dinucleotide metabolic process',\n",
       "  'regulation of meristem growth',\n",
       "  'DNA binding',\n",
       "  'nucleus'))"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# go_description has a many to one relationship with node_id\n",
    "go_description_raw = tuple(frozenset(pd.read_sql(f\"\"\"\n",
    "SELECT node_id, go_description\n",
    "    FROM nodes_raw\n",
    "    WHERE node_id = 'AT4G08920'\n",
    "\"\"\", con)[\"go_description\"]))\n",
    "len(go_description_raw), go_description_raw[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0e56a40a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15, ('GO:0071705', 'GO:0099503', 'GO:0015210', 'GO:0009504', 'GO:0005783'))"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# go_terms has a many to one relationship with transcript\n",
    "go_terms_trsx_raw = tuple(frozenset(pd.read_sql(f\"\"\"\n",
    "SELECT transcript, go_terms\n",
    "    FROM nodes_raw\n",
    "    WHERE transcript = 'AT1G26440.2'\n",
    "\"\"\", con)[\"go_terms\"]))\n",
    "len(go_terms_trsx_raw), go_terms_trsx_raw[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "44bf501a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(6, ('2.3.4.3', '4.2.7.2.3', None, '35.1', '2.2.1.3'))"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mapman_bin has a many to one relationship with node_id\n",
    "mapman_bin_raw = tuple(frozenset(pd.read_sql(f\"\"\"\n",
    "SELECT node_id, mapman_bin\n",
    "    FROM nodes_raw\n",
    "    WHERE node_id = 'AT3G17240'\n",
    "\"\"\", con)[\"mapman_bin\"]))\n",
    "len(mapman_bin_raw), mapman_bin_raw[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "03547a55",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5, ('13.1.1.2.4', '18.4.3.1.4', '15.3.5.6.2.2', None, '14.6.3.1.3'))"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# mapman_bin has a many to one relationship with transcript\n",
    "mapman_bin_trsx_raw = tuple(frozenset(pd.read_sql(f\"\"\"\n",
    "SELECT node_id, mapman_bin\n",
    "    FROM nodes_raw\n",
    "    WHERE transcript = 'AT1G18040.1'\n",
    "\"\"\", con)[\"mapman_bin\"]))\n",
    "len(mapman_bin_trsx_raw), mapman_bin_trsx_raw[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "238b34a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>MAX(pheno_reference)</th>\n",
       "      <th>MIN(pheno_reference)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>As2</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>As75</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>At2</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>Co59</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Mo98</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>Na23</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>SDV</td>\n",
       "      <td>https://doi.org/10.1038/nature08800</td>\n",
       "      <td>doi:10.1038/nature08800</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  node_id                 MAX(pheno_reference)     MIN(pheno_reference)\n",
       "0     As2  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800\n",
       "1    As75  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800\n",
       "2     At2  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800\n",
       "3    Co59  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800\n",
       "4    Mo98  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800\n",
       "5    Na23  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800\n",
       "6     SDV  https://doi.org/10.1038/nature08800  doi:10.1038/nature08800"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pheno_reference differs only in the URL, so keep the URL version\n",
    "pheno_reference_node_ids = '\", \"'.join(list(pd.read_sql(\"\"\"\n",
    "SELECT node_id, pheno_reference\n",
    "    FROM nodes_raw\n",
    "    GROUP BY node_id\n",
    "    HAVING COUNT(DISTINCT(pheno_reference)) = 2\n",
    "    ORDER BY COUNT(DISTINCT(pheno_reference)) DESC\n",
    "\"\"\", con)[\"node_id\"]))\n",
    "pd.read_sql(f\"\"\"\n",
    "SELECT node_id, MAX(pheno_reference), MIN(pheno_reference)\n",
    "    FROM nodes_raw\n",
    "    WHERE node_id in (\"{pheno_reference_node_ids}\")\n",
    "    GROUP BY node_id\n",
    "\"\"\", con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "aa2a5921-1967-4228-b0fb-21b87d467fed",
   "metadata": {},
   "outputs": [],
   "source": [
    "concat_query = \"\"\"\n",
    "SELECT\n",
    "        MIN(\"index\") as mindex,\n",
    "        node_id,\n",
    "        node_type,\n",
    "        gene_model_type,\n",
    "        gene_symbol,\n",
    "        pheno_aragwas_id,\n",
    "        pheno_description,\n",
    "        pheno_pto_description,\n",
    "        pheno_pto_name,\n",
    "        MAX(pheno_reference),\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(transcript)>0\n",
    "            THEN printf('\"%s\"', transcript)\n",
    "            ELSE '\"\"' END\n",
    "        )\n",
    "        || ']' as transcripts,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(go_terms)>0\n",
    "            THEN printf('\"%s\"', go_terms)\n",
    "            ELSE '\"\"' END\n",
    "        )\n",
    "        || ']' as go_terms,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(go_terms)>0\n",
    "            THEN printf(\n",
    "                '{\"transcript\": \"%s\", \"go_term\": \"%s\", \"go_description\": \"%s\"}',\n",
    "                transcript, go_terms, go_description\n",
    "            )\n",
    "            ELSE '{}' END\n",
    "        )\n",
    "        || ']' as go_infos,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(mapman_bin)>0\n",
    "            THEN printf(\n",
    "                '{\"transcript\": \"%s\", \"bin\": \"%s\", \"name\": \"%s\", \"description\": \"%s\"}',\n",
    "                transcript, mapman_bin, mapman_name, mapman_description\n",
    "            )\n",
    "            ELSE '{}' END\n",
    "        )\n",
    "        || ']' as mapman_infos,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(tair_short_description)>0\n",
    "            THEN printf(\n",
    "                '{\n",
    "                    \"transcript\": \"%s\",\n",
    "                    \"gene_model_type\": \"%s\",\n",
    "                    \"tair_short_description\": \"%s\",\n",
    "                    \"tair_curator_summary\": \"%s\",\n",
    "                    \"tair_computational_description\": \"%s\"\n",
    "                }',\n",
    "                transcript,\n",
    "                gene_model_type,\n",
    "                tair_short_description,\n",
    "                tair_curator_summary,\n",
    "                tair_computational_description\n",
    "            )\n",
    "            ELSE '{}' END\n",
    "        )\n",
    "        || ']' as tair_infos\n",
    "    FROM nodes_raw\n",
    "    GROUP BY node_id, node_type\n",
    "    ORDER BY count(*) DESC\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "4cd13892-7a0a-4ac3-8c51-846d0f26c120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cur_catted = con.execute(f\"\"\"INSERT INTO nodes {concat_query}\"\"\")\n",
    "cur_catted.fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "052b3266-4c48-41bb-872c-2d6dcae9321e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mindex</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>gene_model_type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>pheno_aragwas_id</th>\n",
       "      <th>pheno_description</th>\n",
       "      <th>pheno_pto_description</th>\n",
       "      <th>pheno_pto_name</th>\n",
       "      <th>pheno_reference</th>\n",
       "      <th>transcripts</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>go_infos</th>\n",
       "      <th>mapman_infos</th>\n",
       "      <th>tair_infos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>24604</td>\n",
       "      <td>20700</td>\n",
       "      <td>AT5G49090</td>\n",
       "      <td>gene</td>\n",
       "      <td>transposable_element_gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"AT5G49090.1\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{\\n                    \"transcript\": \"AT5G490...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24605</td>\n",
       "      <td>19958</td>\n",
       "      <td>AT5G49380</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24606</td>\n",
       "      <td>20062</td>\n",
       "      <td>AT5G50645</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24607</td>\n",
       "      <td>20326</td>\n",
       "      <td>AT5G53588</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24608</td>\n",
       "      <td>20620</td>\n",
       "      <td>AT5G57181</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24609</td>\n",
       "      <td>20633</td>\n",
       "      <td>AT5G57310</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24610</td>\n",
       "      <td>20669</td>\n",
       "      <td>AT5G57730</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24611</td>\n",
       "      <td>20919</td>\n",
       "      <td>AT5G60548</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24612</td>\n",
       "      <td>20976</td>\n",
       "      <td>AT5G61228</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24613</td>\n",
       "      <td>21245</td>\n",
       "      <td>AT5G64342</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "      <td>[{}]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       mindex    node_id node_type            gene_model_type gene_symbol  \\\n",
       "24604   20700  AT5G49090      gene  transposable_element_gene        None   \n",
       "24605   19958  AT5G49380      gene                       None        None   \n",
       "24606   20062  AT5G50645      gene                       None        None   \n",
       "24607   20326  AT5G53588      gene                       None        None   \n",
       "24608   20620  AT5G57181      gene                       None        None   \n",
       "24609   20633  AT5G57310      gene                       None        None   \n",
       "24610   20669  AT5G57730      gene                       None        None   \n",
       "24611   20919  AT5G60548      gene                       None        None   \n",
       "24612   20976  AT5G61228      gene                       None        None   \n",
       "24613   21245  AT5G64342      gene                       None        None   \n",
       "\n",
       "      pheno_aragwas_id pheno_description pheno_pto_description pheno_pto_name  \\\n",
       "24604             None              None                  None           None   \n",
       "24605             None              None                  None           None   \n",
       "24606             None              None                  None           None   \n",
       "24607             None              None                  None           None   \n",
       "24608             None              None                  None           None   \n",
       "24609             None              None                  None           None   \n",
       "24610             None              None                  None           None   \n",
       "24611             None              None                  None           None   \n",
       "24612             None              None                  None           None   \n",
       "24613             None              None                  None           None   \n",
       "\n",
       "      pheno_reference      transcripts go_terms go_infos mapman_infos  \\\n",
       "24604            None  [\"AT5G49090.1\"]     [\"\"]     [{}]         [{}]   \n",
       "24605            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24606            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24607            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24608            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24609            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24610            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24611            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24612            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "24613            None             [\"\"]     [\"\"]     [{}]         [{}]   \n",
       "\n",
       "                                              tair_infos  \n",
       "24604  [{\\n                    \"transcript\": \"AT5G490...  \n",
       "24605                                               [{}]  \n",
       "24606                                               [{}]  \n",
       "24607                                               [{}]  \n",
       "24608                                               [{}]  \n",
       "24609                                               [{}]  \n",
       "24610                                               [{}]  \n",
       "24611                                               [{}]  \n",
       "24612                                               [{}]  \n",
       "24613                                               [{}]  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "catted = pd.read_sql(\"\"\"SELECT * from nodes\"\"\", con)\n",
    "catted.tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "7facb47b-e1e7-43ad-9d0a-4685585e3d71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, ('AT1G07890.8', 'AT1G07890.1'))"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts = frozenset([transcript for transcript in json.loads(catted[\"transcripts\"].values[0]) if transcript])\n",
    "len(transcripts), tuple(transcripts)[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "golden-teens",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'transcript': 'AT1G07890.5',\n",
       "  'bin': '10.4.1.1',\n",
       "  'name': '.Redox homeostasis.hydrogen peroxide removal.ascorbate-glutathione cycle.ascorbate peroxidase (APX)',\n",
       "  'description': 'ascorbate peroxidase (APX) (original description: pep chromosome:TAIR10:1:2437478:2439665:1 gene:AT1G07890 transcript:AT1G07890.5 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:APX1 description:MEE6 [Source:UniProtKB/TrEMBL;Acc:A0A178W5I1])'},\n",
       " {'transcript': 'AT1G07890.3',\n",
       "  'bin': '10.4.1.1',\n",
       "  'name': '.Redox homeostasis.hydrogen peroxide removal.ascorbate-glutathione cycle.ascorbate peroxidase (APX)',\n",
       "  'description': 'ascorbate peroxidase (APX) (original description: pep chromosome:TAIR10:1:2437725:2439665:1 gene:AT1G07890 transcript:AT1G07890.3 gene_biotype:protein_coding transcript_biotype:protein_coding gene_symbol:APX1 description:MEE6 [Source:UniProtKB/TrEMBL;Acc:A0A178W5I1])'}]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# note that the mapman data is tied to the transcript.\n",
    "[mapman for mapman in json.loads(catted[\"mapman_infos\"].values[0]) if mapman != {}][:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "49ceba42-250a-4608-8548-834340534af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# note that the go term data is tied to the transcript.\n",
    "# there appears to be a one to many relationship from transcripts to GO terms\n",
    "go_infos_json = [json.dumps(go_info) for go_info in json.loads(catted[\"go_infos\"].values[0]) if go_info != {}]\n",
    "assert len(go_infos_json) == len(frozenset(go_infos_json))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "01f2bcc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('GO:0009699', 'GO:0005576', 'GO:0009809', 'GO:0005737', 'GO:0016688')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the list of go terms per gene is also useful to have available\n",
    "go_terms = tuple(frozenset([go_term for go_term in json.loads(catted[\"go_terms\"].values[0]) if go_term != \"\"]))\n",
    "go_terms[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "bf85c8cb-a423-48a1-98b8-d942b9854373",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8,\n",
       " 8,\n",
       " ('{\"transcript\": \"AT1G07890.6\", \"gene_model_type\": \"protein_coding\", \"tair_short_description\": \"ascorbate peroxidase 1\", \"tair_curator_summary\": \"Encodes a cytosolic ascorbate peroxidase APX1.  Ascorbate peroxidases are enzymes that scavenge hydrogen peroxide in plant cells.  Eight types of APX have been described for Arabidopsis: three cytosolic (APX1, APX2, APX6), two chloroplastic types (stromal sAPX, thylakoid tAPX), and three microsomal (APX3, APX4, APX5) isoforms. At least part of the induction of heat shock proteins during light stress in Arabidopsis is mediated by H2O2 that is scavenged by APX1. Expression of the gene is downregulated in the presence of paraquat, an inducer of photoxidative stress. The mRNA is cell-to-cell mobile.\", \"tair_computational_description\": \"ascorbate peroxidase 1;(source:Araport11)\"}',\n",
       "  '{\"transcript\": \"AT1G07890.7\", \"gene_model_type\": \"protein_coding\", \"tair_short_description\": \"ascorbate peroxidase 1\", \"tair_curator_summary\": \"Encodes a cytosolic ascorbate peroxidase APX1.  Ascorbate peroxidases are enzymes that scavenge hydrogen peroxide in plant cells.  Eight types of APX have been described for Arabidopsis: three cytosolic (APX1, APX2, APX6), two chloroplastic types (stromal sAPX, thylakoid tAPX), and three microsomal (APX3, APX4, APX5) isoforms. At least part of the induction of heat shock proteins during light stress in Arabidopsis is mediated by H2O2 that is scavenged by APX1. Expression of the gene is downregulated in the presence of paraquat, an inducer of photoxidative stress. The mRNA is cell-to-cell mobile.\", \"tair_computational_description\": \"ascorbate peroxidase 1;(source:Araport11)\"}'))"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# note that the tair data is also tied to the transcript.\n",
    "# there appears to be a one to one relationship from transcript to tair info\n",
    "tair_infos_json = frozenset([json.dumps(tair_info) for tair_info in json.loads(catted[\"tair_infos\"].values[0]) if tair_info != {}])\n",
    "len(tair_infos_json), len(transcripts), tuple(frozenset(tair_infos_json))[:2]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f5b432c-d241-47fd-9b03-780a5029ee3c",
   "metadata": {},
   "source": [
    "***\n",
    "Did this actually solve the problem? That is, `ATCG00690` is not in the merged nodes table csv."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "42a55174-2203-4ac5-a35e-cdd03602b23b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mindex</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>gene_model_type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>pheno_aragwas_id</th>\n",
       "      <th>pheno_description</th>\n",
       "      <th>pheno_pto_description</th>\n",
       "      <th>pheno_pto_name</th>\n",
       "      <th>pheno_reference</th>\n",
       "      <th>transcripts</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>go_infos</th>\n",
       "      <th>mapman_infos</th>\n",
       "      <th>tair_infos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>ATCG00690</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"ATCG00690.1\",\"ATCG00690.1\",\"ATCG00690.1\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"GO:0009535\",\"GO:0009539\",\"GO:0016168\",...</td>\n",
       "      <td>[{},{},{\"transcript\": \"ATCG00690.1\", \"go_term\"...</td>\n",
       "      <td>[{\"transcript\": \"ATCG00690.1\", \"bin\": \"1.1.1.2...</td>\n",
       "      <td>[{},{\\n                    \"transcript\": \"ATCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mindex    node_id node_type gene_model_type gene_symbol pheno_aragwas_id  \\\n",
       "0      34  ATCG00690      gene            None        None             None   \n",
       "\n",
       "  pheno_description pheno_pto_description pheno_pto_name pheno_reference  \\\n",
       "0              None                  None           None            None   \n",
       "\n",
       "                                         transcripts  \\\n",
       "0  [\"ATCG00690.1\",\"ATCG00690.1\",\"ATCG00690.1\",\"\",...   \n",
       "\n",
       "                                            go_terms  \\\n",
       "0  [\"\",\"\",\"GO:0009535\",\"GO:0009539\",\"GO:0016168\",...   \n",
       "\n",
       "                                            go_infos  \\\n",
       "0  [{},{},{\"transcript\": \"ATCG00690.1\", \"go_term\"...   \n",
       "\n",
       "                                        mapman_infos  \\\n",
       "0  [{\"transcript\": \"ATCG00690.1\", \"bin\": \"1.1.1.2...   \n",
       "\n",
       "                                          tair_infos  \n",
       "0  [{},{\\n                    \"transcript\": \"ATCG...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "problem_gene = pd.read_sql(\"\"\"SELECT * from nodes where node_id=\"ATCG00690\" \"\"\", con)\n",
    "problem_gene"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b62c58bd-3b91-458d-b488-bcdf74476efe",
   "metadata": {},
   "source": [
    "Yes. Yes it did. *phew*"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e7196b59-268f-45b4-96da-ad063d21dab6",
   "metadata": {},
   "source": [
    "***\n",
    "Below `node_id` is used, but in these cases it always represents a gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a0640259-b41e-4673-a68f-4e21f002de98",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mindex</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>tair_computational_description</th>\n",
       "      <th>tair_curator_summary</th>\n",
       "      <th>tair_short_description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>588</td>\n",
       "      <td>AT1G07890</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"ascorbate peroxidase...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"Encodes a cytosolic ...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"ascorbate peroxidase...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>6672</td>\n",
       "      <td>AT4G30820</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>1662</td>\n",
       "      <td>AT1G20620</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"catalase 3;(source:Araport...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"Catalase, catalyzes the br...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"catalase 3\",\"catalase 3\",\"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>11000</td>\n",
       "      <td>AT4G20260</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"plasma-membran...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"Encodes a Ca2+...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"plasma-membran...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>178</td>\n",
       "      <td>AT1G51270</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24609</td>\n",
       "      <td>20633</td>\n",
       "      <td>AT5G57310</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24610</td>\n",
       "      <td>20669</td>\n",
       "      <td>AT5G57730</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24611</td>\n",
       "      <td>20919</td>\n",
       "      <td>AT5G60548</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24612</td>\n",
       "      <td>20976</td>\n",
       "      <td>AT5G61228</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>24613</td>\n",
       "      <td>21245</td>\n",
       "      <td>AT5G64342</td>\n",
       "      <td>gene</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "      <td>[\"\"]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24614 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       mindex    node_id node_type  \\\n",
       "0         588  AT1G07890      gene   \n",
       "1        6672  AT4G30820      gene   \n",
       "2        1662  AT1G20620      gene   \n",
       "3       11000  AT4G20260      gene   \n",
       "4         178  AT1G51270      gene   \n",
       "...       ...        ...       ...   \n",
       "24609   20633  AT5G57310      gene   \n",
       "24610   20669  AT5G57730      gene   \n",
       "24611   20919  AT5G60548      gene   \n",
       "24612   20976  AT5G61228      gene   \n",
       "24613   21245  AT5G64342      gene   \n",
       "\n",
       "                          tair_computational_description  \\\n",
       "0      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"ascorbate peroxidase...   \n",
       "1      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...   \n",
       "2      [\"\",\"\",\"\",\"\",\"\",\"\",\"catalase 3;(source:Araport...   \n",
       "3      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"plasma-membran...   \n",
       "4      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...   \n",
       "...                                                  ...   \n",
       "24609                                               [\"\"]   \n",
       "24610                                               [\"\"]   \n",
       "24611                                               [\"\"]   \n",
       "24612                                               [\"\"]   \n",
       "24613                                               [\"\"]   \n",
       "\n",
       "                                    tair_curator_summary  \\\n",
       "0      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"Encodes a cytosolic ...   \n",
       "1      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...   \n",
       "2      [\"\",\"\",\"\",\"\",\"\",\"\",\"Catalase, catalyzes the br...   \n",
       "3      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"Encodes a Ca2+...   \n",
       "4      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...   \n",
       "...                                                  ...   \n",
       "24609                                               [\"\"]   \n",
       "24610                                               [\"\"]   \n",
       "24611                                               [\"\"]   \n",
       "24612                                               [\"\"]   \n",
       "24613                                               [\"\"]   \n",
       "\n",
       "                                  tair_short_description  \n",
       "0      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"ascorbate peroxidase...  \n",
       "1      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...  \n",
       "2      [\"\",\"\",\"\",\"\",\"\",\"\",\"catalase 3\",\"catalase 3\",\"...  \n",
       "3      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"plasma-membran...  \n",
       "4      [\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",\"\",...  \n",
       "...                                                  ...  \n",
       "24609                                               [\"\"]  \n",
       "24610                                               [\"\"]  \n",
       "24611                                               [\"\"]  \n",
       "24612                                               [\"\"]  \n",
       "24613                                               [\"\"]  \n",
       "\n",
       "[24614 rows x 6 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inspect_query = \"\"\"SELECT\n",
    "        MIN(\"index\") as mindex,\n",
    "        node_id,\n",
    "        node_type,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(tair_computational_description)>0\n",
    "            THEN printf('\"%s\"', tair_computational_description)\n",
    "            ELSE '\"\"' END\n",
    "        )\n",
    "        || ']' as tair_computational_description,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(tair_curator_summary)>0\n",
    "            THEN printf('\"%s\"', tair_curator_summary)\n",
    "            ELSE '\"\"' END\n",
    "        )\n",
    "        || ']' as tair_curator_summary,\n",
    "        '['\n",
    "        || group_concat(\n",
    "            CASE WHEN length(tair_short_description)>0\n",
    "            THEN printf('\"%s\"', tair_short_description)\n",
    "            ELSE '\"\"' END\n",
    "        )\n",
    "        || ']' as tair_short_description\n",
    "    FROM nodes_raw\n",
    "    GROUP BY node_id, node_type\n",
    "    ORDER BY count(*) DESC\n",
    "\"\"\"\n",
    "inspect = pd.read_sql(inspect_query, con)\n",
    "inspect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "controversial-scanning",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ascorbate peroxidase 1;(source:Araport11)'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# is it the case that there is only one nonempty tair_computational_description for a given node_id?\n",
    "distinct_tair_computational_descriptions = tuple(frozenset([tcd for tcd in json.loads(inspect[\"tair_computational_description\"].values[0]) if tcd != \"\"]))\n",
    "assert len(distinct_tair_computational_descriptions) == 1\n",
    "distinct_tair_computational_descriptions[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "62835f02-7207-4540-957f-5b0149008fa7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ascorbate peroxidase 1'"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# is it the case that there is only one nonempty tair_short_description for a given node_id?\n",
    "distinct_tair_short_descriptions = tuple(frozenset([tcs for tcs in json.loads(inspect[\"tair_short_description\"].values[0]) if tcs != \"\"]))\n",
    "assert len(distinct_tair_short_descriptions) == 1\n",
    "distinct_tair_short_descriptions[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "reduced-least",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Encodes a cytosolic ascorbate peroxidase APX1.  Ascorbate peroxidases are enzymes that scavenge hydrogen peroxide in plant cells.  Eight types of APX have been described for Arabidopsis: three cytosolic (APX1, APX2, APX6), two chloroplastic types (stromal sAPX, thylakoid tAPX), and three microsomal (APX3, APX4, APX5) isoforms. At least part of the induction of heat shock proteins during light stress in Arabidopsis is mediated by H2O2 that is scavenged by APX1. Expression of the gene is downregulated in the presence of paraquat, an inducer of photoxidative stress. The mRNA is cell-to-cell mobile.'"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# is it the case that there is only one nonempty tair_curator_summary for a given node_id?\n",
    "distinct_tair_curator_summarys = tuple(frozenset([tcs for tcs in json.loads(inspect[\"tair_curator_summary\"].values[0]) if tcs != \"\"]))\n",
    "assert len(distinct_tair_curator_summarys) == 1\n",
    "distinct_tair_curator_summarys[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "hazardous-house",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AT1G15750</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     node_id  counts\n",
       "0  AT1G15750       3"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# no, that is not the case\n",
    "pd.read_sql(\"\"\"\n",
    "SELECT node_id,\n",
    "        COUNT(DISTINCT(tair_curator_summary)) as counts\n",
    "    FROM nodes_raw\n",
    "    GROUP BY node_id\n",
    "    ORDER BY COUNT(DISTINCT(tair_curator_summary)) DESC, node_id\n",
    "\"\"\", con).head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "direct-explanation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('Encodes a protein with several WD40 repeats at the C-terminus and predicted protein-protein interaction domains at the N-terminus.  Together with the TOPLESS-RELATED PROTEINS (TPRs), it is thought to be involved in transcriptional repression of root-promoting genes in the top help of the embryo during the transition stage of embryogenesis. The ability of IAA12 to repress transcription is diminished in a tpl-1 mutant background.',\n",
       " 'Encodes a protein with several WD40 repeats at the C-terminus and predicted protein-protein interaction domains at the N-terminus.  Together with the TOPLESS-RELATED PROTEINS (TPRs), it is thought to be involved in transcriptional repression of root-promoting genes in the top help of the embryo during the transition stage of embryogenesis.  The ability of IAA12 to repress transcription is diminished in a tpl-1 mutant background.',\n",
       " 'Encodes a protein with several WD40 repeats at the C-terminus and predicted protein-protein interaction domains at the N-terminus.  Together with the TOPLESS-RELATED PROTEINS (TPRs), it is thought to be involved in transcriptional repression of root-promoting genes in the top half of the embryo during the transition stage of embryogenesis. It can also interact with IAA12 through the EAR domain of IAA12 and the CTLH domain of TPL. The ability of IAA12 to repress transcription is diminished in a tpl-1 mutant background.',\n",
       " '---')"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some summaries differ, although some in trivial ways, for instance\n",
    "# \"embryogeneis.\" is followed by one and two spaces below\n",
    "# map null to the value '---'\n",
    "valsdf = pd.read_sql(\"\"\"\n",
    "SELECT node_id,\n",
    "        coalesce(tair_curator_summary, '---') as 'values'\n",
    "    FROM nodes_raw\n",
    "    WHERE node_id='AT1G15750'\n",
    "    GROUP BY coalesce(tair_curator_summary, '---')\n",
    "    ORDER BY node_id DESC\n",
    "\"\"\", con)\n",
    "tuple(valsdf[\"values\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f5a6144",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "08fdd733-acd4-4652-974a-df5959dbdbf9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'AT1G18590', 'AT2G34420', 'AT5G05850', 'ATCG00280', 'ATCG00350', 'ATCG00680', 'ATCG00690'\n"
     ]
    }
   ],
   "source": [
    "print(f\"\"\"'{\"', '\".join(sorted(('AT1G18590', 'AT2G34420', 'AT5G05850', 'ATCG00280', 'ATCG00350', 'ATCG00680', 'ATCG00690')))}'\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "cc9c1f62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 checks, 5 results\n"
     ]
    }
   ],
   "source": [
    "node_checks = ('AT1G18590', 'AT2G34420', 'AT5G05850', 'ATCG00280', 'ATCG00350', 'ATCG00680', 'ATCG00690')\n",
    "checks = pd.read_sql(f\"\"\"SELECT * from nodes where node_id in (\"{'\", \"'.join(node_checks)}\")\"\"\", con)\n",
    "print(f\"{len(node_checks)} checks, {len(checks.values)} results\")\n",
    "# assert len(node_checks) == len(checks.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "20baf20f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mindex</th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>gene_model_type</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>pheno_aragwas_id</th>\n",
       "      <th>pheno_description</th>\n",
       "      <th>pheno_pto_description</th>\n",
       "      <th>pheno_pto_name</th>\n",
       "      <th>pheno_reference</th>\n",
       "      <th>transcripts</th>\n",
       "      <th>go_terms</th>\n",
       "      <th>go_infos</th>\n",
       "      <th>mapman_infos</th>\n",
       "      <th>tair_infos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>ATCG00280</td>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PSBC</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"ATCG00280.1\",\"ATCG00280.1\",\"ATCG00280.1\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"GO:0009507\",\"GO:0009539\",\"GO:0003729\",...</td>\n",
       "      <td>[{},{},{\"transcript\": \"ATCG00280.1\", \"go_term\"...</td>\n",
       "      <td>[{\"transcript\": \"ATCG00280.1\", \"bin\": \"1.1.1.2...</td>\n",
       "      <td>[{},{\\n                    \"transcript\": \"ATCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>AT2G34420</td>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>LHCB1.5</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"AT2G34420.1\",\"AT2G34420.1\",\"AT2G34420.1\",\"AT...</td>\n",
       "      <td>[\"\",\"\",\"GO:0009534\",\"GO:0009535\",\"GO:0009535\",...</td>\n",
       "      <td>[{},{},{\"transcript\": \"AT2G34420.1\", \"go_term\"...</td>\n",
       "      <td>[{\"transcript\": \"AT2G34420.1\", \"bin\": \"1.1.1.1...</td>\n",
       "      <td>[{},{\\n                    \"transcript\": \"AT2G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>163</td>\n",
       "      <td>ATCG00350</td>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>PSAA</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"ATCG00350.1\",\"ATCG00350.1\",\"ATCG00350.1\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"GO:0009535\",\"GO:0009536\",\"GO:0009534\",...</td>\n",
       "      <td>[{},{},{\"transcript\": \"ATCG00350.1\", \"go_term\"...</td>\n",
       "      <td>[{\"transcript\": \"ATCG00350.1\", \"bin\": \"1.1.4.2...</td>\n",
       "      <td>[{},{\\n                    \"transcript\": \"ATCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>1496</td>\n",
       "      <td>AT1G18590</td>\n",
       "      <td>gene</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SOT17</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"AT1G18590.1\",\"AT1G18590.1\",\"\",\"\",\"AT1G18590....</td>\n",
       "      <td>[\"\",\"\",\"GO:0005634\",\"GO:0051923\",\"GO:0005737\",...</td>\n",
       "      <td>[{},{},{\"transcript\": \"\", \"go_term\": \"GO:00056...</td>\n",
       "      <td>[{\"transcript\": \"AT1G18590.1\", \"bin\": \"9.5.1.9...</td>\n",
       "      <td>[{},{\\n                    \"transcript\": \"AT1G...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>34</td>\n",
       "      <td>ATCG00690</td>\n",
       "      <td>gene</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[\"ATCG00690.1\",\"ATCG00690.1\",\"ATCG00690.1\",\"\",...</td>\n",
       "      <td>[\"\",\"\",\"GO:0009535\",\"GO:0009539\",\"GO:0016168\",...</td>\n",
       "      <td>[{},{},{\"transcript\": \"ATCG00690.1\", \"go_term\"...</td>\n",
       "      <td>[{\"transcript\": \"ATCG00690.1\", \"bin\": \"1.1.1.2...</td>\n",
       "      <td>[{},{\\n                    \"transcript\": \"ATCG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mindex    node_id node_type gene_model_type gene_symbol pheno_aragwas_id  \\\n",
       "0      18  ATCG00280      gene  protein_coding        PSBC             None   \n",
       "1       0  AT2G34420      gene  protein_coding     LHCB1.5             None   \n",
       "2     163  ATCG00350      gene  protein_coding        PSAA             None   \n",
       "3    1496  AT1G18590      gene  protein_coding       SOT17             None   \n",
       "4      34  ATCG00690      gene            None        None             None   \n",
       "\n",
       "  pheno_description pheno_pto_description pheno_pto_name pheno_reference  \\\n",
       "0              None                  None           None            None   \n",
       "1              None                  None           None            None   \n",
       "2              None                  None           None            None   \n",
       "3              None                  None           None            None   \n",
       "4              None                  None           None            None   \n",
       "\n",
       "                                         transcripts  \\\n",
       "0  [\"ATCG00280.1\",\"ATCG00280.1\",\"ATCG00280.1\",\"\",...   \n",
       "1  [\"AT2G34420.1\",\"AT2G34420.1\",\"AT2G34420.1\",\"AT...   \n",
       "2  [\"ATCG00350.1\",\"ATCG00350.1\",\"ATCG00350.1\",\"\",...   \n",
       "3  [\"AT1G18590.1\",\"AT1G18590.1\",\"\",\"\",\"AT1G18590....   \n",
       "4  [\"ATCG00690.1\",\"ATCG00690.1\",\"ATCG00690.1\",\"\",...   \n",
       "\n",
       "                                            go_terms  \\\n",
       "0  [\"\",\"\",\"GO:0009507\",\"GO:0009539\",\"GO:0003729\",...   \n",
       "1  [\"\",\"\",\"GO:0009534\",\"GO:0009535\",\"GO:0009535\",...   \n",
       "2  [\"\",\"\",\"GO:0009535\",\"GO:0009536\",\"GO:0009534\",...   \n",
       "3  [\"\",\"\",\"GO:0005634\",\"GO:0051923\",\"GO:0005737\",...   \n",
       "4  [\"\",\"\",\"GO:0009535\",\"GO:0009539\",\"GO:0016168\",...   \n",
       "\n",
       "                                            go_infos  \\\n",
       "0  [{},{},{\"transcript\": \"ATCG00280.1\", \"go_term\"...   \n",
       "1  [{},{},{\"transcript\": \"AT2G34420.1\", \"go_term\"...   \n",
       "2  [{},{},{\"transcript\": \"ATCG00350.1\", \"go_term\"...   \n",
       "3  [{},{},{\"transcript\": \"\", \"go_term\": \"GO:00056...   \n",
       "4  [{},{},{\"transcript\": \"ATCG00690.1\", \"go_term\"...   \n",
       "\n",
       "                                        mapman_infos  \\\n",
       "0  [{\"transcript\": \"ATCG00280.1\", \"bin\": \"1.1.1.2...   \n",
       "1  [{\"transcript\": \"AT2G34420.1\", \"bin\": \"1.1.1.1...   \n",
       "2  [{\"transcript\": \"ATCG00350.1\", \"bin\": \"1.1.4.2...   \n",
       "3  [{\"transcript\": \"AT1G18590.1\", \"bin\": \"9.5.1.9...   \n",
       "4  [{\"transcript\": \"ATCG00690.1\", \"bin\": \"1.1.1.2...   \n",
       "\n",
       "                                          tair_infos  \n",
       "0  [{},{\\n                    \"transcript\": \"ATCG...  \n",
       "1  [{},{\\n                    \"transcript\": \"AT2G...  \n",
       "2  [{},{\\n                    \"transcript\": \"ATCG...  \n",
       "3  [{},{\\n                    \"transcript\": \"AT1G...  \n",
       "4  [{},{\\n                    \"transcript\": \"ATCG...  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "084df4e6",
   "metadata": {},
   "source": [
    "***\n",
    "What about the merged node file?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "96081f15",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>node_id</th>\n",
       "      <th>node_type</th>\n",
       "      <th>transcript</th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>gene_model_type</th>\n",
       "      <th>go_description</th>\n",
       "      <th>pheno_description</th>\n",
       "      <th>pheno_pto_name</th>\n",
       "      <th>pheno_pto_description</th>\n",
       "      <th>pheno_reference</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>21671</td>\n",
       "      <td>ATMG01320</td>\n",
       "      <td>gene</td>\n",
       "      <td>ATMG01320.1</td>\n",
       "      <td>NAD2B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>NADH dehydrogenase (ubiquinone) activity, NADH...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21672</td>\n",
       "      <td>ATMG01350</td>\n",
       "      <td>gene</td>\n",
       "      <td>ATMG01350.1</td>\n",
       "      <td>ORF145C</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21673</td>\n",
       "      <td>ATMG01360</td>\n",
       "      <td>gene</td>\n",
       "      <td>ATMG01360.1</td>\n",
       "      <td>COX1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>cytochrome-c oxidase activity</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21674</td>\n",
       "      <td>ATMG01370</td>\n",
       "      <td>gene</td>\n",
       "      <td>ATMG01370.1</td>\n",
       "      <td>ORF111D</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>21675</td>\n",
       "      <td>ATMG01400</td>\n",
       "      <td>gene</td>\n",
       "      <td>ATMG01400.1</td>\n",
       "      <td>ORF105B</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>molecular_function</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         node_id node_type   transcript gene_symbol gene_model_type  \\\n",
       "21671  ATMG01320      gene  ATMG01320.1       NAD2B  protein_coding   \n",
       "21672  ATMG01350      gene  ATMG01350.1     ORF145C  protein_coding   \n",
       "21673  ATMG01360      gene  ATMG01360.1        COX1  protein_coding   \n",
       "21674  ATMG01370      gene  ATMG01370.1     ORF111D  protein_coding   \n",
       "21675  ATMG01400      gene  ATMG01400.1     ORF105B  protein_coding   \n",
       "\n",
       "                                          go_description pheno_description  \\\n",
       "21671  NADH dehydrogenase (ubiquinone) activity, NADH...               NaN   \n",
       "21672                                 molecular_function               NaN   \n",
       "21673                      cytochrome-c oxidase activity               NaN   \n",
       "21674                                 molecular_function               NaN   \n",
       "21675                                 molecular_function               NaN   \n",
       "\n",
       "      pheno_pto_name pheno_pto_description pheno_reference  \n",
       "21671            NaN                   NaN             NaN  \n",
       "21672            NaN                   NaN             NaN  \n",
       "21673            NaN                   NaN             NaN  \n",
       "21674            NaN                   NaN             NaN  \n",
       "21675            NaN                   NaN             NaN  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge_df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "0bce4483",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(nan, 85),\n",
       " ('AT1G01010.1', 1),\n",
       " ('AT1G01020.6', 1),\n",
       " ('AT1G01030.2', 1),\n",
       " ('AT1G01040.2', 1),\n",
       " ('AT1G01050.2', 1),\n",
       " ('AT1G01060.8', 1),\n",
       " ('AT1G01070.2', 1),\n",
       " ('AT1G01080.3', 1),\n",
       " ('AT1G01090.1', 1)]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(merge_df[\"transcript\"]).most_common(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4938ecf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('As2', 1),\n",
       " ('As75', 1),\n",
       " ('At2', 1),\n",
       " ('Co59', 1),\n",
       " ('Mo98', 1),\n",
       " ('Na23', 1),\n",
       " ('SDV', 1),\n",
       " ('AT1G01010', 1),\n",
       " ('AT1G01020', 1),\n",
       " ('AT1G01030', 1)]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Counter(merge_df[\"node_id\"]).most_common(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9aae4fe7",
   "metadata": {},
   "source": [
    "***\n",
    "I originally thought the following would be useful if mapman data were \"mostly\" 1-1 with gene, but it looks like it is \"mostly\" 1-1 with transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5556d1ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_items(items):\n",
    "    out = dict()\n",
    "    for item in items:\n",
    "        for key, value in item.items():\n",
    "            if key not in out:\n",
    "                out[key] = value\n",
    "                continue\n",
    "            value_old = out[key]\n",
    "            if not isinstance(value_old, list):\n",
    "                out[key] = [value_old, value]\n",
    "                continue\n",
    "            out[key].append(value)\n",
    "    return out\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "c94c85c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'name': ['sookie', 'ming'], 'bells': True}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cat_items = [{\"name\": \"sookie\", \"bells\": True}, {\"name\": \"ming\"}]\n",
    "cat_out = {\"name\" : [\"sookie\", \"ming\"], \"bells\": True}\n",
    "cat_merge = merge_items(cat_items)\n",
    "assert json.dumps(cat_merge) == json.dumps(cat_out)\n",
    "cat_merge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2b4f201",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
